warehouse/migrations/versions/be443e514e3e_data_migration_for_useruniquelogin_ip_.py

```diff
@@ -0,0 +1,101 @@
+# SPDX-License-Identifier: Apache-2.0
+"""
+Data migration for UserUniqueLogin.ip_address_id
+
+Revision ID: be443e514e3e
+Revises: df52c3746740
+Create Date: 2025-12-02 17:32:29.770684
+"""
+
+import sqlalchemy as sa
+
+from alembic import op
+
+
+revision = "be443e514e3e"
+down_revision = "df52c3746740"
+
+
+def _get_remaining_logins_to_update(conn):
+    return conn.execute(
+        sa.text("SELECT COUNT(*) FROM user_unique_logins WHERE ip_address_id IS NULL")
+    ).scalar_one()
+
+
+def _get_remaining_ips_to_insert(conn):
+    return conn.execute(
+        sa.text(
+            """
+            SELECT COUNT(DISTINCT user_unique_logins.ip_address)
+            FROM user_unique_logins
+            LEFT JOIN ip_addresses ON user_unique_logins.ip_address::inet = ip_addresses.ip_address
+            WHERE ip_addresses.id IS NULL AND user_unique_logins.ip_address IS NOT NULL
+            """
+        )
+    ).scalar_one()
+
+
+def upgrade():
+    op.execute("SET statement_timeout = 120000")
+    op.execute("SET lock_timeout = 120000")
+
+    op.create_index(
+        "ix_user_unique_logins_ip_address_migration",
+        "user_unique_logins",
+        ["ip_address"],
+        unique=False,
+    )
+
+    bind = op.get_bind()
+    BATCH_SIZE = 1000
+
+    while _get_remaining_ips_to_insert(bind) > 0:
+        bind.execute(
+            sa.text(
+                """
+                INSERT INTO ip_addresses (ip_address)
```

Comment:
We have WAREHOUSE_IP_SALT in the environment, so we could calculate the hashed_ip_address at the same time. `sha256(CONCAT(ip_address, "WAREHOUSE_IP_SALT")`

This may also surface addresses that already exist in the table.

Reply:
5696067659b12995474a02a0639fa7a4c55a7638